### PR TITLE
Feat/improve tests

### PR DESCRIPTION
**Pull request by parth-15**

### CODE REVIEW

File: src/interfaces/IParaSwapLiquiditySwapAdapter.sol [hunk 1]
```diff
@@ -22,18 +18,18 @@ interface IParaSwapLiquiditySwapAdapter is IBaseParaSwapAdapter {
     uint256 newCollateralAmount; // the minimum amount of new collateral asset to receive
     uint256 offset; // offset in sell calldata in case of swapping all collateral, otherwise 0
     address user; // the address of user
+    bool toFlashloan; // true if flashloan is needed to swap collateral, otherwise false
```
Comment (miguelmtzinf) [inline src/interfaces/IParaSwapLiquiditySwapAdapter.sol:21]:
```suggestion
    bool withFlashLoan; // true if flashloan is needed to swap collateral, otherwise false
```

File: src/interfaces/IParaSwapLiquiditySwapAdapter.sol [hunk 2]
```diff
@@ -22,18 +18,18 @@ interface IParaSwapLiquiditySwapAdapter is IBaseParaSwapAdapter {
...
   /**
    * @notice Swaps liquidity(collateral) from one asset to another
    * @param liquiditySwapParams struct describing the liquidity swap
-   * @param flashParams optional struct describing flashloan params if needed
    * @param collateralATokenPermit optional permit for collateral aToken
    */
   function swapLiquidity(
     LiquiditySwapParams memory liquiditySwapParams,
-    FlashParams memory flashParams,
+    // FlashParams memory flashParams,
```
Comment (miguelmtzinf) [inline src/interfaces/IParaSwapLiquiditySwapAdapter.sol:32]: remove

File: src/interfaces/IParaSwapRepayAdapter.sol [hunk 3]
```diff
@@ -22,19 +18,18 @@ interface IParaSwapRepayAdapter is IBaseParaSwapAdapter {
     uint256 debtRepayAmount; // the amount of debt to repay
     uint256 debtRepayMode; // debt interest rate mode (1 for stable, 2 for variable)
     uint256 offset; // offset in buy calldata in case of swapping all collateral, otherwise 0
+    uint256 flashLoanAmount; // the amount to flashloan equivalent to the debt to repay
```
Comment (miguelmtzinf) [inline src/interfaces/IParaSwapRepayAdapter.sol:21]: there is no `debtRepayAmount` already?

Comment (parth-15) [reply src/interfaces/IParaSwapRepayAdapter.sol:21]: `flashloanAmount` will not be equal to `debtRepayAmount`. Flashloan will be of collateral asset which will be swapped to debt asset. Both will be different.

File: src/contracts/ParaSwapLiquiditySwapAdapter.sol [hunk 4]
```diff
@@ -64,23 +64,21 @@ abstract contract ParaSwapLiquiditySwapAdapter is
...
   function swapLiquidity(
     LiquiditySwapParams memory liquiditySwapParams,
-    FlashParams memory flashParams,
     PermitInput memory collateralATokenPermit
   ) external nonReentrant {
     // Offset in August calldata if wanting to swap all balance, otherwise 0
     if (liquiditySwapParams.offset != 0) {
       (, , address aToken) = _getReserveData(liquiditySwapParams.collateralAsset);
-      uint256 balance = IERC20(aToken).balanceOf(liquiditySwapParams.user);
-      require(balance <= liquiditySwapParams.collateralAmountToSwap, 'INSUFFICIENT_AMOUNT_TO_SWAP');
```
Comment (miguelmtzinf) [inline src/contracts/ParaSwapLiquiditySwapAdapter.sol:74]: this is a sanity check we must keep.
"If the user wants to swap all, the amount used for the API call and passed as argument should be more than the full balance"

File: src/contracts/ParaSwapLiquiditySwapAdapter.sol [hunk 5]
```diff
@@ -64,23 +64,21 @@ abstract contract ParaSwapLiquiditySwapAdapter is
...
-      liquiditySwapParams.collateralAmountToSwap = balance;
+      liquiditySwapParams.collateralAmountToSwap = IERC20(aToken).balanceOf(
+        liquiditySwapParams.user
+      );
     }
 
-    // Non-zero amount if wanting to flashloan, otherwise 0
```
Comment (miguelmtzinf) [inline src/contracts/ParaSwapLiquiditySwapAdapter.sol:78]: change natspec docs, do not remove

File: src/contracts/ParaSwapRepayAdapter.sol [hunk 6]
```diff
@@ -279,9 +276,6 @@ abstract contract ParaSwapRepayAdapter is
     uint256 currentDebt = IERC20(debtToken).balanceOf(user);
 
     if (buyAllBalanceOffset != 0) {
-      // Sanity check to ensure the passed value `debtRepayAmount` is higher than the current debt
-      // when repaying all debt.
-      require(currentDebt <= debtRepayAmount, 'INSUFFICIENT_AMOUNT_TO_REPAY');
```
Comment (miguelmtzinf) [inline src/contracts/ParaSwapRepayAdapter.sol:284]: we want to keep this sanity check too

File: src/contracts/ParaSwapWithdrawSwapAdapter.sol [hunk 7]
```diff
@@ -48,9 +48,7 @@ abstract contract ParaSwapWithdrawSwapAdapter is
 
     // Offset in August calldata if wanting to swap all balance, otherwise 0
     if (withdrawSwapParams.allBalanceOffset != 0) {
-      uint256 balance = IERC20(aToken).balanceOf(withdrawSwapParams.user);
-      require(balance <= withdrawSwapParams.oldAssetAmount, 'INSUFFICIENT_AMOUNT_TO_SWAP');
```
Comment (miguelmtzinf) [inline src/contracts/ParaSwapWithdrawSwapAdapter.sol:52]: same here

File: src/interfaces/IParaSwapRepayAdapter.sol (hunk 3 above)
Comment (miguelmtzinf) [reply src/interfaces/IParaSwapRepayAdapter.sol:21]: that's not `maxCollateralAmountToSwap` then?

File: tests/LiquiditySwapV2.t.sol [hunk 8]
```diff
@@ -90,8 +91,54 @@ contract LiquiditySwapAdapterV2 is BaseTest {
...
+        paraswapData: abi.encode(psp.swapCalldata, psp.augustus)
+      });
+
+    IParaSwapLiquiditySwapAdapter.PermitInput memory collateralATokenPermit;
+
+    vm.expectRevert(bytes(Errors.VL_TRANSFER_NOT_ALLOWED));
+    liquiditySwapAdapter.swapLiquidity(liquiditySwapParams, collateralATokenPermit);
+  }
+
+  function test_revert_wrong_paraswap_route() public {
```
Comment (miguelmtzinf) [inline tests/LiquiditySwapV2.t.sol:117]: add docs explaining why the paraswap route is wrong.

File: tests/LiquiditySwapV2.t.sol [hunk 9]
```diff
@@ -276,64 +388,128 @@ contract LiquiditySwapAdapterV2 is BaseTest {
 
     IERC20Detailed(collateralAssetAToken).approve(
       address(liquiditySwapAdapter),
-      collateralAmountToSwap
+      oldCollateralAssetATokenBalanceBefore
```
Comment (miguelmtzinf) [inline tests/LiquiditySwapV2.t.sol:391]: better to use `collateralAmountToSwap` no? which is the value used for the psp call

File: tests/LiquiditySwapV2.t.sol [hunk 10]
```diff
@@ -276,64 +388,128 @@ contract LiquiditySwapAdapterV2 is BaseTest {
...
+    );
+    uint256 newCollateralAssetATokenBalanceBefore = IERC20Detailed(newCollateralAssetAToken)
+      .balanceOf(user);
+
+    uint256 collateralAmountToSwap = 5000 ether;
+    uint256 expectedAmount = 4800e6;
+    PsPResponse memory psp = _fetchPSPRoute(
+      collateralAsset,
+      newCollateralAsset,
+      (collateralAmountToSwap * 9991) / 10_000, //taking flashloan premium(0.09%) into account
```
Comment (miguelmtzinf) [inline tests/LiquiditySwapV2.t.sol:448]: is this not taken into account in the adapter itself? is this needed?

File: tests/LiquiditySwapV2.t.sol [hunk 11]
```diff
@@ -347,32 +523,82 @@ contract LiquiditySwapAdapterV2 is BaseTest {
...
+    );
+    uint256 newCollateralAssetATokenBalanceBefore = IERC20Detailed(newCollateralAssetAToken)
+      .balanceOf(user);
+
+    uint256 collateralAmountToSwap = 18_000 ether; // supplyAmount
+    uint256 expectedAmount = 17_500e6;
+    PsPResponse memory psp = _fetchPSPRoute(
+      collateralAsset,
+      newCollateralAsset,
+      (collateralAmountToSwap * 9991) / 10_000, //taking flashloan premium(0.09%) into account
```
Comment (miguelmtzinf) [inline tests/LiquiditySwapV2.t.sol:567]: same

File: tests/LiquiditySwapV3.t.sol [hunk 12]
```diff
@@ -280,61 +393,124 @@ contract LiquiditySwapAdapterV3 is BaseTest {
...
         2
       )
     );
-    assertEq(daiCollateralAssetATokenBalanceAfter, 0);
-    _invariant(address(liquiditySwapAdapter), newCollateralAsset, newCollateralAssetAToken);
-    _invariant(address(liquiditySwapAdapter), collateralAsset, collateralAssetAToken);
+    assertGt(
+      newCollateralAssetATokenBalanceAfter - newCollateralAssetATokenBalanceBefore,
+      expectedAmount,
+      'INVALID_AMOUNT_RECEIVED'
```
Comment (miguelmtzinf) [inline tests/LiquiditySwapV3.t.sol:494]: regular msg error here

File: tests/LiquiditySwapV3.t.sol [hunk 13]
```diff
@@ -354,33 +530,82 @@ contract LiquiditySwapAdapterV3 is BaseTest {
...
+      oldCollateralAssetATokenBalanceBefore
+    );
+
+    liquiditySwapAdapter.swapLiquidity(liquiditySwapParams, collateralATokenPermit);
+    uint256 oldCollateralAssetATokenBalanceAfter = IERC20Detailed(collateralAssetAToken).balanceOf(
+      user
+    );
+    uint256 newCollateralAssetATokenBalanceAfter = IERC20Detailed(newCollateralAssetAToken)
+      .balanceOf(user);
+    assertEq(oldCollateralAssetATokenBalanceAfter, 0, 'FULL_SWAP_FAILED');
```
Comment (miguelmtzinf) [inline tests/LiquiditySwapV3.t.sol:601]: same

File: tests/RepayAdapterV2.t.sol [hunk 14]
```diff
@@ -12,9 +12,11 @@ import {AugustusRegistry} from '../src/lib/AugustusRegistry.sol';
 import {IParaSwapRepayAdapter} from '../src/interfaces/IParaSwapRepayAdapter.sol';
 import {stdMath} from 'forge-std/StdMath.sol';
 
-contract RepayAdapterV2 is BaseTest {
+contract RepayAdapterV2Test is BaseTest {
   ParaSwapRepayAdapterV2 internal repayAdapter;
 
+  event ConsoleLog(uint256, uint256);
```
Comment (miguelmtzinf) [inline tests/RepayAdapterV2.t.sol:18]: ?

File: src/script/Deploy.s.sol
Comment (miguelmtzinf) [inline src/script/Deploy.s.sol:1]: rename this file please

File: src/script/Deploy_ParaSwapWithdrawAdapter.s.sol [hunk 15]
```diff
@@ -13,9 +13,9 @@ import {AaveV3Avalanche} from 'aave-address-book/AaveV3Avalanche.sol';
...
 import {AaveV3Arbitrum} from 'aave-address-book/AaveV3Arbitrum.sol';
 import {AaveV3Base} from 'aave-address-book/AaveV3Base.sol';
-import {ParaSwapWithdrawSwapAdapterV2} from '../src/contracts/ParaSwapWithdrawSwapAdapterV2.sol';
-import {ParaSwapWithdrawSwapAdapterV3} from '../src/contracts/ParaSwapWithdrawSwapAdapterV3.sol';
-import {AugustusRegistry} from '../src/lib/AugustusRegistry.sol';
+import {ParaSwapWithdrawSwapAdapterV2} from 'src/contracts/ParaSwapWithdrawSwapAdapterV2.sol';
+import {ParaSwapWithdrawSwapAdapterV3} from 'src/contracts/ParaSwapWithdrawSwapAdapterV3.sol';
+import {AugustusRegistry} from 'src/lib/AugustusRegistry.sol';
 
 contract EthereumV2 is EthereumScript {
```
Comment (miguelmtzinf) [inline src/script/Deploy_ParaSwapWithdrawAdapter.s.sol:20]: update the address of the owner in all contracts please. You should point to AaveGovV3 Level 1 Executor

File: tests/BuyAdapterFuzz.t.sol [hunk 16]
```diff
@@ -0,0 +1,87 @@
...
+      AugustusRegistry.ETHEREUM
+    );
+    aaveV3EthereumAssets = [
+      AaveV3EthereumAssets.DAI_UNDERLYING,
+      AaveV3EthereumAssets.LINK_UNDERLYING,
+      AaveV3EthereumAssets.LUSD_UNDERLYING
+    ];
+  }
+
+  function test_fuzz_buy_on_paraswap(
```
Comment (miguelmtzinf) [inline tests/BuyAdapterFuzz.t.sol:31]:
```suggestion
  function test_fuzz_buyOnParaSwap(
```

File: tests/mocks/MockParaSwapBuyAdapter.sol
Comment (miguelmtzinf) [inline tests/mocks/MockParaSwapBuyAdapter.sol:1]: rename to harness instead of mock

File: tests/BuyAdapterFuzz.t.sol [hunk 17]
```diff
@@ -0,0 +1,87 @@
...
+    address assetToSwapTo = aaveV3EthereumAssets[toAssetIndex];
+    PsPResponse memory psp = _fetchPSPRouteWithoutPspCacheUpdate(
+      assetToSwapFrom,
+      assetToSwapTo,
+      amountToBuy,
+      user,
+      false,
+      swapAll
+    );
+    if (swapAll) {
```
Comment (miguelmtzinf) [inline tests/BuyAdapterFuzz.t.sol:54]: why is this needed?

File: tests/BuyAdapterFuzz.t.sol [hunk 18]
```diff
@@ -0,0 +1,87 @@
...
+      // In memory, swapCalldata consists of a 256 bit length field, followed by
+      // the actual bytes data, that is why 32 is added to the byte offset.
+      assembly {
+        amountAtOffset := mload(add(swapCalldata, add(offset, 32)))
+      }
+      assertEq(amountAtOffset, amountToBuy, 'wrong offset');
+    }
+    deal(assetToSwapFrom, address(buyAdapter), psp.srcAmount);
+
+    buyAdapter.buyOnParaSwap(
```
Comment (miguelmtzinf) [inline tests/BuyAdapterFuzz.t.sol:71]: add check for event and `assetToSwapFrom`

File: tests/BuyAdapterFuzz.t.sol [hunk 19]
```diff
@@ -0,0 +1,87 @@
...
+  function setUp() public override {
+    super.setUp();
+    vm.createSelectFork(vm.rpcUrl('mainnet'));
+
+    buyAdapter = new MockParaSwapBuyAdapter(
+      IPoolAddressesProvider(address(AaveV3Ethereum.POOL_ADDRESSES_PROVIDER)),
+      address(AaveV3Ethereum.POOL),
+      AugustusRegistry.ETHEREUM
+    );
+    aaveV3EthereumAssets = [
```
Comment (miguelmtzinf) [inline tests/BuyAdapterFuzz.t.sol:24]: can we use the `getReserves` function of the pool so we are not limited to only 3 assets?

File: tests/BuyAdapterFuzz.t.sol [hunk 20]
```diff
@@ -0,0 +1,87 @@
...
+    uint256 amountToBuy,
+    bool swapAll
+  ) public {
+    uint256 totalAssets = aaveV3EthereumAssets.length;
+    fromAssetIndex = bound(fromAssetIndex, 0, totalAssets - 1);
+    toAssetIndex = bound(toAssetIndex, 0, totalAssets - 1);
+    if (fromAssetIndex == toAssetIndex) {
+      toAssetIndex = (toAssetIndex + 1) % totalAssets;
+    }
+    amountToBuy = bound(amountToBuy, 1e9, 1000 ether);
```
Comment (miguelmtzinf) [inline tests/BuyAdapterFuzz.t.sol:43]: can we increase the upper bound here?

File: tests/scripts/psp.js [hunk 21]
```diff
@@ -24,6 +24,7 @@ const FROM_DECIMALS = Number(args[8]);
 const TO_DECIMALS = Number(args[9]);
 // param only needed for the hash
 // const BLOCK_NUMBER = Number(args[10]);
+const UPDATE_PSP_CACHE = args[11] !== "false";
```
Comment (miguelmtzinf) [inline tests/scripts/psp.js:27]: `USE_CACHE` sounds better

File: tests/utils/BaseTest.sol [hunk 22]
```diff
@@ -57,6 +57,34 @@ contract BaseTest is Test {
     return abi.decode(res, (PsPResponse));
   }
 
+  function _fetchPSPRouteWithoutPspCacheUpdate(
```
Comment (miguelmtzinf) [inline tests/utils/BaseTest.sol:60]: why we do not refactor and merge these 2 functions into a single 1

File: tests/PSRouteFuzz.t.sol [hunk 23]
```diff
@@ -0,0 +1,44 @@
...
+
+  function test_fuzz_correct_offset(uint256 fromAmount, bool sell) public {
+    fromAmount = bound(fromAmount, 1e9, 1_000_000 ether);
+    address fromAsset = AaveV3EthereumAssets.DAI_UNDERLYING;
+    address toAsset = AaveV3EthereumAssets.LUSD_UNDERLYING;
+
+    PsPResponse memory psp = _fetchPSPRouteWithoutPspCacheUpdate(
+      fromAsset,
+      toAsset,
+      fromAmount,
```
Comment (miguelmtzinf) [inline tests/PSRouteFuzz.t.sol:23]: this is just `amount` right? `fromAmount` or `toAmount` depending on direction

File: tests/PSRouteFuzz.t.sol [hunk 24]
```diff
@@ -0,0 +1,44 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import {IERC20Detailed} from '@aave/core-v3/contracts/dependencies/openzeppelin/contracts/IERC20Detailed.sol';
+import {AaveV3EthereumAssets} from 'aave-address-book/AaveV3Ethereum.sol';
+import {BaseTest} from './utils/BaseTest.sol';
+import {stdMath} from 'forge-std/StdMath.sol';
```
Comment (miguelmtzinf) [inline tests/PSRouteFuzz.t.sol:7]:
```suggestion
import {AaveV3EthereumAssets} from 'aave-address-book/AaveV3Ethereum.sol';
import {BaseTest} from './utils/BaseTest.sol';
```

File: tests/SellAdapterFuzz.t.sol [hunk 25]
```diff
@@ -0,0 +1,92 @@
...
+      AugustusRegistry.ETHEREUM
+    );
+    aaveV3EthereumAssets = [
+      AaveV3EthereumAssets.DAI_UNDERLYING,
+      AaveV3EthereumAssets.LINK_UNDERLYING,
+      AaveV3EthereumAssets.LUSD_UNDERLYING
+    ];
+  }
+
+  function test_fuzz_sell_on_paraswap(
```
Comment (miguelmtzinf) [inline tests/SellAdapterFuzz.t.sol:31]:
```suggestion
  function test_fuzz_sellOnParaSwap(
```

File: tests/SellAdapterFuzz.t.sol [hunk 26]
```diff
@@ -0,0 +1,92 @@
...
+    uint256 amountToSwap,
+    bool swapAll
+  ) public {
+    uint256 totalAssets = aaveV3EthereumAssets.length;
+    fromAssetIndex = bound(fromAssetIndex, 0, totalAssets - 1);
+    toAssetIndex = bound(toAssetIndex, 0, totalAssets - 1);
+    if (fromAssetIndex == toAssetIndex) {
+      toAssetIndex = (toAssetIndex + 1) % totalAssets;
+    }
+    amountToSwap = bound(amountToSwap, 1e9, 1000 ether);
```
Comment (miguelmtzinf) [inline tests/SellAdapterFuzz.t.sol:43]: same

File: README.md [hunk 27]
```diff
@@ -324,7 +306,7 @@ Security considerations around the ParaSwap adapter contracts:
...
     - Example for `BUY`: API call of `buy(x,y)` and swap transaction of `buy(x',y')`. If `y'>y` then `x'>x`, so positive slippage happens as user is receiving more assets than expected.
     - Example for `SELL`: API call of `sell(x,y)` and swap transaction of `sell(x',y')`. If `x'>x` then `y'>y`, so positive slippage happens as user is receiving more assets than expected.
-
+  - In ParaSwapLiquditySwapAdapter, fetching of ParaSwap route for swapping from collateral asset to another asset with flashloan enabled should take flashloanFee into consideration. As ParaSwapLiquditySwapAdapter swaps (collateralAmountToSwap - flashloanFee) to guarantee that flashloanFee is paid, generating routes with (collateralAmountToSwap - flashloanFee) is recommended.
```
Comment (miguelmtzinf) [inline README.md:309]: Add an example please.

File: yarn.lock
Comment (miguelmtzinf) [inline yarn.lock:1]: remove this file, and add the package-lock instead please

File: tests/utils/BaseTest.sol [hunk 28]
```diff
@@ -69,6 +97,27 @@ contract BaseTest is Test {
     );
   }
 
+  /**
+   * @dev Ensure offset is correct when swapping full amount
+   */
+  function _ensureCorrectOffset(
```
Comment (miguelmtzinf) [inline tests/utils/BaseTest.sol:103]:
```suggestion
  /**
   * @dev Ensure the amount offset in ParaSwap calldata is correct
   */
  function _checkAmountInParaSwapCalldata(
```

File: README.md [hunk 29]
```diff
@@ -100,29 +111,203 @@ struct PermitInput {
...
+    - Example for `BUY`: API call of `buy(x,y)` and swap transaction of `buy(x',y')`. If `y'>y` then `x'>x`, so positive slippage happens as user is receiving more assets than expected.
+    - Example for `SELL`: API call of `sell(x,y)` and swap transaction of `sell(x',y')`. If `x'>x` then `y'>y`, so positive slippage happens as user is receiving more assets than expected.
+  - In ParaSwapLiquditySwapAdapter, fetching of ParaSwap route for swapping from collateral asset to another asset with flashloan enabled should take flashloanFee into consideration. As ParaSwapLiquditySwapAdapter swaps (collateralAmountToSwap - flashloanFee) to guarantee that flashloanFee is paid, generating routes with (collateralAmountToSwap - flashloanFee) is recommended.
```
Comment (miguelmtzinf) [inline README.md:309]:
```suggestion
  - In `ParaSwapLiquditySwapAdapter`, fetching of ParaSwap route for swapping from collateral asset to another asset with flashloan enabled should take `flashloanFee` into consideration. As `ParaSwapLiquditySwapAdapter` swaps `(collateralAmountToSwap - flashloanFee)` to guarantee that `flashloanFee` is paid, generating routes with `(collateralAmountToSwap - flashloanFee) `is recommended.
```

File: README.md [hunk 30]
```diff
@@ -100,29 +111,203 @@ struct PermitInput {
...
+    - Example: User wants to swap `1000 BUSD` collateral to `min(995 USDC)` collateral and with flashloan enabled. The ParaSwap route should be generated for selling (1000 BUSD - 0.9 BUSD) to buy USDC assuming (0.09% flashloan fee). Thus, ParaSwapLiquditySwapAdapter will flashloan (1000 BUSD) but will sell (1000 BUSD - 0.9 BUSD) to ensure that 0.9 BUSD stays in the contract to pay flashloan premium.
```
Comment (miguelmtzinf) [inline README.md:310]:
```suggestion
    - Example: User wants to swap `1000 BUSD` collateral to `min(995 USDC)` collateral and with flashloan enabled. The ParaSwap route should be generated for selling `(1000 BUSD - 0.9 BUSD)` to buy USDC assuming (0.09% flashloan fee). Thus, `ParaSwapLiquditySwapAdapter` will flashloan `(1000 BUSD)` but will sell `(1000 BUSD - 0.9 BUSD)` to ensure that `0.9 BUSD` stays in the contract to pay flashloan premium.
```

**Review by miguelmtzinf — APPROVED** [review]
